File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<int>> &vis, vector<vector<char>>& grid, int drow[], int dcol[]){
        vis[row][col] = 1;
        int n = grid.size();
        int m = grid[0].size();

        for(int i = 0;i < 4; i++){
            int nrow = row + drow[i];
            int ncol = col + dcol[i];

            if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && !vis[nrow][ncol] && grid[nrow][ncol] == '1'){
                dfs(nrow, ncol, vis, grid, drow, dcol);
            }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));
        int count = 0;
        int dcol[] = {0, 1, 0, -1};
        int drow[] = {-1, 0, 1, 0};

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(!vis[i][j] && grid[i][j] == '1'){
                    count++;
                    dfs(i, j, vis, grid, drow, dcol);
                }
            }
        }
        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def dfs(self, row, col, grid, vis, drow, dcol):
        vis[row][col] = 1
        n = len(grid)
        m = len(grid[0])

        for i in range(4):
            nrow = row + drow[i]
            ncol = col + dcol[i]

            if nrow >=0 and nrow < n and ncol >= 0 and ncol < m and vis[nrow][ncol] == 0 and grid[nrow][ncol] == "1":
                self.dfs(nrow, ncol, grid, vis, drow, dcol)

    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])
        vis = [[0] * m for _ in range(n)]
        drow = [-1, 0, 1, 0]
        dcol = [0, 1, 0, -1]
        count = 0

        for i in range(n):
            for j in range(m):
                if vis[i][j] == 0 and grid[i][j] == "1":
                    count += 1
                    self.dfs(i, j, grid, vis, drow, dcol)
        
        return count